builder/evaluator.go

```diff
@@ -308,7 +311,14 @@ func (b *Builder) dispatch(stepN int, ast *parser.Node) error {
 		str = ast.Value
 		if _, ok := replaceEnvAllowed[cmd]; ok {
 			var err error
-			str, err = ProcessWord(ast.Value, b.Config.Env)
+			// Append the build-time environment to config-environment.
+			// This allows builder config to override the variables, making the behavior similar to
+			// a shell script i.e. `ENV foo bar` overrides value of `foo` passed in build
+			// context. But `ENV foo $foo` will use the value from build context if one
+			// isn't already been defined by a previous ENV primitive.
+			envs := b.Config.Env
```

Comment:
Just as an optimization, can we check to see if its safe to move these two
`envs =...` lines to before the for loop (line 305). Then we only do the `append()`
once and not for each ast node.  I think this is safe because we're not
recursive within one dockerfile command.  But hopefully our testing will let
use know for sure.